crates/chia-client/src/peer.rs

```diff
@@ -1,363 +1,341 @@
-
-        // Send the message.
-        let bytes = match stream(&message) {
-            Ok(bytes) => bytes.into(),
-            Err(error) => {
-                self.requests.lock().await.remove(&message_id);
-                return Err(error.into());
-            }
-        };
-        let send_result = self.sink.lock().await.send(bytes).await;
-
-        if let Err(error) = send_result {
-            self.requests.lock().await.remove(&message_id);
-            return Err(error.into());
+            id: Some(self.0.requests.insert(sender).await),
+            data: body.to_bytes()?.into(),
         }
+        .to_bytes()?
+        .into();
 
-        // Wait for the response.
-        let response = receiver.await;
-
-        // Remove the one shot channel.
-        self.requests.lock().await.remove(&message_id);
-
-        // Handle the response, if present.
-        response.or(Err(Error::MissingResponse))
+        self.0.sink.lock().await.send(message).await?;
+        Ok(receiver.await?)
     }
+}
 
-    pub fn receiver(&self) -> &broadcast::Receiver<PeerEvent> {
-        &self.event_receiver
+impl Drop for PeerInner {
+    fn drop(&mut self) {
+        self.inbound_handle.abort();
     }
+}
 
-    pub fn receiver_mut(&mut self) -> &mut broadcast::Receiver<PeerEvent> {
-        &mut self.event_receiver
-    }
+async fn handle_inbound_messages(
+    mut stream: Stream,
+    sender: mpsc::Sender<Message>,
+    requests: Arc<RequestMap>,
+) -> Result<()> {
+    use tungstenite::Message::{Binary, Close, Frame, Ping, Pong, Text};
 
-    async fn handle_inbound(
-        message: WsMessage,
-        requests: &Requests,
-        event_sender: &broadcast::Sender<PeerEvent>,
-    ) -> Result<(), Error<()>> {
-        // Parse the message.
-        let message = Message::from_bytes(message.into_data().as_ref())?;
-
-        if let Some(id) = message.id {
-            // Send response through oneshot channel if present.
-            if let Some(request) = requests.lock().await.remove(&id) {
-                request.send(message).ok();
-            }
-            return Ok(());
-        }
+    while let Some(message) = stream.next().await {
+        let message = message?;
 
-        macro_rules! events {
-            ( $( $event:ident ),+ $(,)? ) => {
-                match message.msg_type {
-                    $( ProtocolMessageTypes::$event => {
-                        event_sender
-                            .send(PeerEvent::$event($event::from_bytes(message.data.as_ref())?))
-                            .ok();
-                    } )+
-                    _ => {}
-                }
-            };
+        match message {
+            Text(text) => {
+                log::warn!("Received unexpected text message: {text}");
+            }
+            Close(close) => {
+                log::warn!("Received close: {close:?}");
+                break;
+            }
+            Ping(_ping) => {}
```

Comment:
are we not expected to (explicitly) respond with a pong in this case?

Reply:
Tungstenite (the WebSocket library) handles pings automatically, so I'm not sure why this is even a message type. I don't think it's ever received.